solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+	{
+		Diagnostic diag{};
+		diag.message = "Hello, FIXME's should be fixed.";
+		diag.range.start = _file.buffer().toPosition(pos);
+		diag.range.end = {diag.range.start.line, diag.range.start.column + 5};
+		diag.severity = DiagnosticSeverity::Error;
+		diag.source = "solc";
+		params.diagnostics.emplace_back(diag);
+	}
+
+	for (size_t pos = _file.contentString().find("TODO", 0); pos != string::npos; pos = _file.contentString().find("FIXME", pos + 1))
+	{
+		Diagnostic diag{};
+		diag.message = "Please remember to create a ticket on GitHub for that.";
+		diag.range.start = _file.buffer().toPosition(pos);
+		diag.range.end = {diag.range.start.line, diag.range.start.column + 5};
+		diag.severity = DiagnosticSeverity::Hint;
+		diag.source = "solc";
+		params.diagnostics.emplace_back(diag);
+	}
+#endif
+
+	_result.emplace_back(params);
+}
+
+frontend::ASTNode const* LanguageServer::findASTNode(lsp::Position const& _position, std::string const& _fileName)
+{
+	if (!m_compilerStack)
+		return nullptr;
+
+	frontend::ASTNode const& sourceUnit = m_compilerStack->ast(_fileName);
+	auto const sourcePos = sourceUnit.location().source->translateLineColumnToPosition(_position.line + 1, _position.column + 1);
+
+	ASTNodeLocator m{sourcePos};
+	sourceUnit.accept(m);
+	auto const closestMatch = m.closestMatch();
+	if (closestMatch != nullptr)
+		fprintf(stderr, "findASTNode for %s @ pos=%d:%d (%d), symbol: '%s' (%s)\n",
+				sourceUnit.location().source->name().c_str(),
+				sourcePos,
+				_position.line,
+				_position.column,
+				closestMatch->location().text().c_str(), typeid(*closestMatch).name());
+	else
+		fprintf(stderr, "findASTNode for pos=%d:%d (%d), not found.\n",
+				sourcePos,
+				_position.line,
+				_position.column);
+
+	return closestMatch;
+}
+
+optional<LanguageServer::Location> LanguageServer::gotoDefinition(DocumentPosition _location)
+{
+	if (auto const file = m_vfs.find(_location.uri); file != nullptr)
+	{
+		compile(*file);
+		solAssert(m_compilerStack.get() != nullptr, "");
+
+		auto const sourceName = file->uri().substr(7); // strip "file://"
+
+		if (auto const sourceNode = findASTNode(_location.position, sourceName); sourceNode)
+		{
+			if (auto const importDirective = dynamic_cast<ImportDirective const*>(sourceNode))
+			{
+				// When cursor is on an import directive, then we want to jump to the actual file that
+				// is being imported.
+				if (auto const fpm = m_fileReader->fullPathMapping().find(importDirective->path()); fpm != m_fileReader->fullPathMapping().end())
+				{
+					Location output{};
+					output.uri = "file://" + fpm->second;
+					return {output};
+				}
+				else
+					return nullopt; // definition not found
+			}
+			else if (auto const declaration = dynamic_cast<Declaration const*>(sourceNode))
+			{
+				// For any kind of declaration, jump to the referencing symbol of that declaration.
```

Comment:
Can you explain how this works in practice?

Reply:
Not sure how detailed, hmm...

* The client sends a `textDocument/definition` (go to definition) request, containing a cursor position. 
* I'm looking for the corresponding AST node that matches this cursor position in this file
* if that AST node is having an annotated declaration it will respond with the location to the name of that declaration.

Now, while doing this rubber-duck talking, I realize I can delete this if-branch. It's useless  code (it does something, but it's meaningless). Thanks. :-)